src/main/host/descriptor/socket/inet/legacy_tcp.rs

```diff
@@ -946,28 +946,66 @@ impl LegacyTcpSocket {
 
-        // sanity check: make sure new socket peer address matches address returned from
-        // tcp_acceptServerPeer() above
+        // Associate the new socket with the local:peer address pair. Previously the new socket
+        // was never registered this way; instead the packets continued to be routed to the parent
+        // listening-socket, which was responsible for routing them to this new "child" socket.
+        // But that led to bugs such as https://github.com/shadow/shadow/issues/3563.
         {
             let File::Socket(Socket::Inet(InetSocket::LegacyTcp(new_socket))) =
                 open_file.inner_file()
             else {
                 panic!("Expected this to be a LegacyTcpSocket");
             };
 
-            let new_socket = new_socket.borrow();
+            let new_socket_ref = new_socket.borrow();
 
-            let mut ip: libc::in_addr_t = 0;
-            let mut port: libc::in_port_t = 0;
+            // sanity check: make sure new socket peer address matches address returned from
+            // tcp_acceptServerPeer() above
+            {
+                let mut ip: libc::in_addr_t = 0;
+                let mut port: libc::in_port_t = 0;
 
-            // should return ip and port in network byte order
-            let okay = unsafe {
-                c::legacysocket_getPeerName(new_socket.as_legacy_socket(), &mut ip, &mut port)
-            };
+                // should return ip and port in network byte order
+                let okay = unsafe {
+                    c::legacysocket_getPeerName(
+                        new_socket_ref.as_legacy_socket(),
+                        &mut ip,
+                        &mut port,
+                    )
+                };
+
+                assert_eq!(okay, 1);
+                assert_eq!(ip, peer_addr.sin_addr.s_addr);
+                assert_eq!(port, peer_addr.sin_port);
+            }
 
-            assert_eq!(okay, 1);
-            assert_eq!(ip, peer_addr.sin_addr.s_addr);
-            assert_eq!(port, peer_addr.sin_port);
+            let peer_addr = new_socket_ref
+                .getpeername()
+                .expect("error finding peer address")
+                .expect("missing peer address");
+            let peer_addr = SocketAddrV4::from(peer_addr);
+
+            let local_addr = new_socket_ref
+                .getsockname()
+                .expect("error finding local address")
+                .expect("missing local address");
+            let local_addr = SocketAddrV4::from(local_addr);
+
+            let (_addr, handle) = inet::associate_socket(
+                InetSocket::LegacyTcp(Arc::clone(new_socket)),
+                local_addr,
+                peer_addr,
+                /* Allow the parent/listening socket to be bound to the same address,
+                 * with a missing/generic peer. */
+                /* check_generic_peer= */
+                false,
+                net_ns,
+                rng,
+            )?;
```

Comment:
Something is unclear for me here. Is the local address that is associated with the new socket here using the same ip/port as the listening socket? I think before it was, and I think it still should be, but now I can't tell if we've changed that behavior from the diff.

The reason we had the awkward listening parent / child relationship was to make sure the server port does not change after the connection is accepted. So if a client connects to a server at `1.1.1.1:443`, then from the client's perspective all packets coming from the server on that connection should come from `1.1.1.1:443`, even if the server uses a new (child) socket to handle it after calling `accept()`. I just want to make sure this diff is not changing that, e.g., changing the port `443` to some other port.

Reply:
Added some assertions to `LegacyTcpSocket::accept` to validate expectations about the local address of the child socket.

`_tcpchild_new` was already setting the child local address to be that of the parent, *but* that included setting the address to unspecified if the listening socket address was unspecified (i.e. bound to all interfaces).

I'm looking at adding a test for this but it doesn't fit neatly into `test_accept.rs` since the bind helper functions always bind to `INADDR_LOOPBACK`. It probably makes sense to extend them to also test ~everything when binding to `INADDR_ANY` instead of `INADDR_LOOPBACK`. I'll poke at that a bit more, but I might file an issue instead of climbing down that rabbit hole right now.

Reply:
> but that included setting the address to unspecified if the listening socket address was unspecified (i.e. bound to all interfaces).

Forgot to say I fixed this in the just-pushed squash commit. The child socket now gets bound to the address on which it received the connection

Reply:
Yeah that seems right to me (it should keep the 4-tuple consistent). Thanks!